pkg/view/router.go

```diff
@@ -1,74 +1,386 @@
+			log.Err(err).Msg("failed to get session")
+			return
+		}
+		session.AddFlash(Flash{
+			Type:    ty,
+			Message: m,
+		})
+		err = session.Save(r, w)
+		if err != nil {
+			log.Err(err).Msg("failed to save session")
 		}
 	}
 
 	r.Get("/", func(w http.ResponseWriter, r *http.Request) {
-		err := gv.Render(w, http.StatusOK, "pages/index", loadModel(r))
+		err := gv.Render(w, http.StatusOK, "pages/index", loadModel(r, w))
 		if err != nil {
 			http.Error(w, err.Error(), http.StatusInternalServerError)
 		}
 	})
 
 	r.Get("/connections", func(w http.ResponseWriter, r *http.Request) {
-		err := gv.Render(w, http.StatusOK, "pages/connections/index", loadModel(r))
+		user, ok := getUser(r)
+		if !ok {
+			http.Error(w, "User not found", http.StatusInternalServerError)
+			return
+		}
+		destinations, err := storageServices.Database.GetDestinations(r.Context(), user.ID)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		m := loadModel(r, w)
+		m.Connections = Connections{
+			Destinations: destinations,
+		}
+		err = gv.Render(w, http.StatusOK, "pages/connections/index", m)
 		if err != nil {
 			http.Error(w, err.Error(), http.StatusInternalServerError)
 		}
 	})
 
 	r.Get("/connections/new", func(w http.ResponseWriter, r *http.Request) {
-		err := gv.Render(w, http.StatusOK, "pages/connections/new", loadModel(r))
+		err := gv.Render(w, http.StatusOK, "pages/connections/upsert", loadModel(r, w))
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+	})
+
+	r.Post("/connections/upsert", func(w http.ResponseWriter, r *http.Request) {
+		user, ok := getUser(r)
+		if !ok {
+			http.Error(w, "User not found", http.StatusInternalServerError)
+			return
+		}
+
+		err := r.ParseForm()
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		id := r.Form.Get("id")
+		if id != "" {
+			http.Error(w, "Editing connections not yet supported", http.StatusBadRequest)
+			return
+		}
+
+		settings := map[string]any{}
+		name := r.Form.Get("name")
+
+		t := r.Form.Get("type")
+		switch t {
+		case "duckdb":
+			// XXX breadchris support file destination?
+
+			tok := r.Form.Get("token")
+			if tok == "" {
+				http.Error(w, "Must specify a token", http.StatusBadRequest)
+				return
+			}
+			db := r.Form.Get("database")
+			if db == "" {
+				http.Error(w, "Must specify a token", http.StatusBadRequest)
+				return
+			}
+
+			// XXX breadchris what validation should be done here?
+
+			settings["token"] = tok
+			settings["database"] = db
+		default:
+			http.Error(w, "Unknown connection type", http.StatusBadRequest)
+			return
+		}
+
+		cd := config.Destination{
+			Type:     t,
+			Settings: settings,
+		}
+
+		err = destManager.TestCredentials(cd)
+		if err != nil {
+			log.Err(err).Msg("failed to connect to destination")
+			newFlash(w, r, FlashTypeError, "Failed to connect to destination. Check the settings and try again.")
```

Comment:
I'm fine with returning the error string from line 253 since it will have a more descriptive error.